server/api/sample_sizes.py

```diff
@@ -41,7 +41,9 @@ def sample_sizes_for_contest(contest: Contest) -> dict:
             cumulative_results,
         )
 
-    targeted_contests = Contest.query.filter_by(is_targeted=True)
+    targeted_contests = Contest.query.filter_by(
+        election_id=election.id, is_targeted=True
+    )
```

Comment:
i think this was just a bug... yikes

Reply:
good catch. Let's add some tests that have other elections in there just hanging around to make sure we're truly not joining against them, ever.

Reply:
So, I've been thinking about this problem and it's feeling kind of tricky. In order to do this well, we need to make sure that *every* query filters by election, which means we either need to write a new test for every query, or add the extra elections to the db for every test. Writing a new test for every query seems like a lot of work, and also a practice we'd likely fail to keep up. Adding extra elections to the db for every test seems easier and more future-proof.

However, even adding an extra election to the db in this case wouldn't necessarily make the test fail. The extra election would have to have a targeted contest that yielded a larger sample size than the one the test is already using - in other words, the data design has to have knowledge of the logic to reliably cause a failure. We could design a test for this bug in particular, but I'm not confident this will make for a robust general approach for unknown future bugs.

That makes me think about potentially building in some sort of check into the SQLAlchemy models to ensure that every query for a model includes the election_id field. For models that don't currently have an election_id, we could either add one (good for sharding in the future, though prob not necessary) or we could make those models check for a different foreign key that has the same logical effect (e.g. for a Batch, check that jurisdiction_id is included, since Jurisdictions are election-specific).

Thoughts?

This would have been a pretty gnarly bug and we got pretty damn lucky to catch it so I'm definitely inclined to try to do something to prevent it from happening again.

Reply:
I agree 100% with your point, that if we can build something that forces a where clause by election ID, we should. Very much worth the investment. Every query should have one, except where there is an explicit opt out named with an appropriately scary name.